### PR TITLE
Add table to store federation relationships

### LINKED ISSUE #2261: Federation Management API: datastore changes
Design and implement the datastore changes needed to persist the changes done through the Federation Management API.

### PR DESCRIPTION
**Pull request by amartinezfayo**
This PR adds a table in the datastore to store federation relationships with foreign trust domains.

The table has the trust domain name and the information needed to get updated bundles of that trust domain from a SPIFFE bundle endpoint server.

Fixes #2261.

### CODE REVIEW

File: pkg/server/datastore/sqlstore/models.go [hunk 1]
```diff
@@ -123,6 +123,33 @@ func (DNSName) TableName() string {
 	return "dns_names"
 }
 
+// FederatedTrustDomain holds federated trust domains.
+// It has the information needed to get updated bundles of the
+// federated trust domain from a SPIFFE bundle endpoint server.
+type FederatedTrustDomain struct {
```
Comment (azdagron) [inline pkg/server/datastore/sqlstore/models.go:129]: @evan2645 has mused before about a way to designate a federated trust domain as "federated with" by default (from an entry point of view. If we're confident that we want to implement such a feature, we might want a column for that...

Comment (amartinezfayo) [reply pkg/server/datastore/sqlstore/models.go:129]: That sounds good! I'm adding a new bool field `federate_with_new_registration_entries` for this purpose.

File: pkg/server/datastore/sqlstore/models.go [hunk 2]
```diff
@@ -123,6 +123,37 @@ func (DNSName) TableName() string {
...
+	BundleEndpointProfile string
+
+	// EndpointSPIFFEID specifies the expected SPIFFE ID of the
+	// SPIFFE bundle endpoint server when BundleEndpointProfile
+	// is "https_spiffe"
+	EndpointSPIFFEID string
+
+	// FederateWithNewRegistrationEntries indicates wether the trust domain
+	// automatically federates with new registration entries by default or not.
+	FederateWithNewRegistrationEntries bool
```
Comment (azdagron) [inline pkg/server/datastore/sqlstore/models.go:149]: This implies that existing registration entries won't federate with this trust domain. @evan2645 can correct me if I'm wrong, but I _think_ the behavior we want is that if this boolean is set, there is an implied `FederatesWith` for _all_ registration entries. I think the implementation of this feature would not involve actually changing the FederatesWith list for any entries when they are stored, but rather adding these trust domains to the list on egress.
If so, maybe a better name would be something simple like `Implicit bool`.

Comment (amartinezfayo) [reply pkg/server/datastore/sqlstore/models.go]: I was not sure if the intention was to include existing registration entries in this behavior or not. I had some doubts about it but now that I think it more it makes sense.
I renamed the field to the suggested `implicit` name, which looks a lot better :)

**Review by azdagron — APPROVED** [review]
\o/. LGTM!

**Comment by azdagron** [comment]
Merged with admin. We think we have a corrupt linter cache being picked up by this job somehow.

**Pull request merged**